### PR TITLE
[MRG] add masked_batch_dot layer, and its tests and docs

### PR DESCRIPTION
**Pull request by nelson-liu**
Added a layer for doing `masked_batch_dot`.

Built Docs:
- API Summary page for layers: https://132-47793051-gh.circle-artifacts.com/0/home/ubuntu/deep_qa/doc/_build/html/layers/backend.html#maskedbatchdot
- Full API Reference page for layers: https://132-47793051-gh.circle-artifacts.com/0/home/ubuntu/deep_qa/doc/_build/html/api_doc/deep_qa.layers.backend.html#module-deep_qa.layers.backend.masked_batch_dot

### CODE REVIEW

File: src/main/python/deep_qa/layers/backend/masked_batch_dot.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        - a_dot_b
+    """
+    def __init__(self, **kwargs):
+        self.supports_masking = True
+        super(MaskedBatchDot, self).__init__(**kwargs)
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        # We do not need a mask beyond this layer.
```
Comment (matt-gardner) [inline src/main/python/deep_qa/layers/backend/masked_batch_dot.py]: The fact that you're not returning a mask should be in the class docstring, including some kind of statement saying that `MaskedBatchDot` returns zeros in places that were masked.  If this then gets fed into a softmax, for instance, your computation will be incorrect.

File: src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        tensor_a = Input(shape=(3, 2), dtype='float32')
+        tensor_b = Input(shape=(4, 2), dtype='float32')
+        masked_batch_dot = MaskedBatchDot()([tensor_a, tensor_b])
+        model = Model(input=[tensor_a, tensor_b], output=[masked_batch_dot])
+
+        tensor_a = numpy.array([[[0.3, 0.1], [0.4, 0.2], [0.8, 0.1]],
+                                [[0.3, 0.1], [0.4, 0.2], [0.8, 0.1]]])
+        tensor_b = numpy.array([[[0.2, 0.6], [0.4, 0.3], [0.5, 0.7], [0.1, .6]],
+                                [[0.2, 0.6], [0.4, 0.3], [0.5, 0.7], [0.1, .6]]])
+        adotb = model.predict([tensor_a, tensor_b])
```
Comment (matt-gardner) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: `a_dot_b` would be nicer here.

File: src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        assert adotb.shape == (batch_size, 3, 4)
+        assert_almost_equal(adotb, numpy.array([[[ 0.12, 0.15000001, 0.22, 0.09],
+                                                 [ 0.20000002, 0.22000001, 0.34, 0.16000001],
+                                                 [ 0.22000001, 0.35000002, 0.47, 0.14000002]],
+                                                [[ 0.12, 0.15000001, 0.22, 0.09],
+                                                 [ 0.20000002, 0.22000001, 0.34, 0.16000001],
+                                                 [ 0.22000001, 0.35000002, 0.47, 0.14000002]]]))
+
```
Comment (matt-gardner) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: I know it's a bit of a pain, but can you add a test for a case where there's a mask, too?  You should just be able to test the `call()` method directly with `K.eval()`, and that's enough.

File: src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    def test_get_output_shape_for(self):
+        # TODO(nelson): verify the correctness of this, esp. in rank 1 edge case.
```
Comment (nelson-liu) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: @matt-gardner not sure if i'm computing the output shape correctly, could you give me explicit verification?

Comment (matt-gardner) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: I'm not sure what you mean by "rank 1 edge case".  The shapes looked correct to me.

Comment (nelson-liu) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: I based this off of the docs for `K.batch_dot`, and they had a special case when the final output is rank one but I'm not too sure if i captured it.

Comment (matt-gardner) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: To capture that, you'd want something like `assert mbd.get_output_shape_for([(5, 10), (5, 10)]) == (5, 1)`

Comment (nelson-liu) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: i see, i wasn't capturing that in my code before so i've updated the test and fixed it.

File: src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                                                   [0.2, 0.22, 0.34, 0.16],
+                                                   [0.22, 0.35, 0.47, 0.14]],
+                                                  [[0.12, 0.15, 0.22, 0.09],
+                                                   [0.2, 0.22, 0.34, 0.16],
+                                                   [0.22, 0.35, 0.47, 0.14]]]))
+
+    def test_call_works_on_masked_input(self):
+        tensor_a = K.variable(numpy.array([[[0.3, 0.1], [0.4, 0.2], [0.8, 0.1]],
+                                           [[0.3, 0.1], [0.4, 0.2], [0.8, 0.1]]]))
+        print(tensor_a)
```
Comment (nelson-liu) [inline src/test/python/deep_qa_test/layers/backend/masked_batch_dot_test.py]: reminder to remove this later

**Comment by nelson-liu** [comment]
I don't think this is really needed anymore, as it's mostly superseded by #191. closing, let me know if you want to add it.
